### PR TITLE
hm2-CRUD-REST-SERVICE

### PR DESCRIPTION
**Pull request by shuspes**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,42 @@
+{
+  "name": "nodejs-mentoring-api",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "build": "npx babel src --out-dir dist",
+    "dev": "nodemon --exec npm run restart",
+    "restart": "rm -rf dist && npm run build && npm run start",
+    "start": "node dist/app.js",
```
Comment (andrei-akula-epam) [inline package.json]: For non production code you can just use `babel-node`
https://babeljs.io/docs/en/babel-node
e.g.
`"start": "nodemon --exec babel-node index.js"`

File: src/app.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+app.get('/', (req, res) => {
+    console.log('root is called');
+    res.send('Hi from root');
+});
+
+// eslint-disable-next-line no-unused-vars
+app.use((err, req, res, next) => {
+    console.error('ERROR: ', err);
+    res.status(500).send({ 'error': err.message });
+});
```
Comment (andrei-akula-epam) [inline src/app.js]: for The default error handler please look at https://expressjs.com/en/guide/error-handling.html#the-default-error-handler
You should insert  
`if (res.headersSent) {
    return next(err)
  }`
before sending 500.

File: src/controllers/users.controller.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import databaseWrapper from '../databaseWrapper';
+
+export default class UsersController {
+    #validator = createValidator({ passError: true });
+
+    #userBodySchema = Joi.object({
+        'login': Joi.string().regex(/^[a-zA-Z0-9-]*$/).required(), // only letters, numbers and -
+        'password': Joi.string().regex(/^(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).*$/).required(), // password must contain at least one letter, one big letter and number
+        'age': Joi.number().min(4).max(130).required(), // user’s age must be between 4 and 130.
+        'isDeleted': Joi.boolean().required()
```
Comment (andrei-akula-epam) [inline src/controllers/users.controller.js]: I guess this field is not part of either POST request body nor PUT request body.
It's internal property.

File: src/controllers/users.controller.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    getUser = (req, res) => {
+        const userId = req.params.userId || '';
+        const user = this.databaseWrapper.getUser(userId);
+        res.send({ user });
+    }
+
+    createUser = (req, res) => {
+        const userFields = req.body;
+        const user = this.databaseWrapper.createUser(userFields);
+        res.send({ user });
```
Comment (andrei-akula-epam) [inline src/controllers/users.controller.js]: By default HTTP status code on a response is 200.
In this case, If POST request succeeded in creating a new user you should respond with 201 status code.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

File: src/routes/users.route.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+router.get('/:loginSubstring/:limit', usersController.getAutoSuggestUsers);
+
+router.delete('/:userId', usersController.removeUser);
+
+router.use(usersController.handleUserBodyValidationMiddleware);
+
+// eslint-disable-next-line no-unused-vars
+router.use((err, req, res, next) => {
+    console.error('ERROR: ', err);
+    res.status(400).send({ 'error': err.message });
```
Comment (andrei-akula-epam) [inline src/routes/users.route.js]: You don't filter errors. I mean you don't differentiate between your own errors (thrown by  UserDatabaseWrapper ) and possible errors from third party code.

File: src/routes/users.route.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+const router = express.Router();
+const usersController = new UsersController();
+
+router.use(express.json());
+
+router.param('userId', usersController.handleUserIdParamMiddleware);
+
+router.get('/', usersController.getAllUsers);
+
+router.get('/:userId', usersController.getUser);
```
Comment (andrei-akula-epam) [inline src/routes/users.route.js]: Why don't you use `router.route` ?
You could write 
```javascript
router.route('/:userId')
  .get(usersController.getUser)
  .put(...)
  .delete(...);
```

File: src/routes/users.route.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
+import express from 'express';
+import UsersController from '../controllers/users.controller';
+
+const router = express.Router();
+const usersController = new UsersController();
+
+router.use(express.json());
```
Comment (andrei-akula-epam) [inline src/routes/users.route.js:8]: This line is usually put in server index file, `src/app.js` in your case.
`router.use(express.json());` applies middleware for all routes, not only specific one.

File: README.md [hunk 8]
```diff
@@ -1 +1,31 @@
...
+
+* **GET** get user by id - curl -i http://localhost:3000/users/ac34d1e0-8ffe-11eb-a8b3-0242ac130001 | npx json
+
+* **POST** create user - curl -i -X POST -H "Content-Type: application/json" -d "{\"login\": \"user-5-new\", \"password\": \"P@ssw0rd\", \"age\": 56, \"isDeleted\": false}" http://localhost:3000/users | npx json
+
+* **PUT** update user - curl -i -X PUT -H "Content-Type: application/json" -d "{\"login\": \"user-2-updated\", \"password\": \"U$er@1\", \"age\": 56, \"isDeleted\": false}" http://localhost:3000/users/ac34d83e-8ffe-11eb-a8b3-0242ac130002 | npx json
+
+* **GET** get auto suggest users - curl -i http://localhost:3000/users/use/10 | npx json
+
+* **DELETE** remive user - curl -i -X DELETE http://localhost:3000/users/ac34d1e0-8ffe-11eb-a8b3-0242ac130001 | npx json
```
Comment (andrei-akula-epam) [inline README.md]: `remove` typo

Comment (shuspes) [reply README.md:21]: thanks! fixed

Comment (shuspes) [reply README.md]: > `remove` typo

thanks! fixed

File: package.json (hunk 1 above)
Comment (shuspes) [reply package.json]: good point, thanks. done

File: src/controllers/users.controller.js (hunk 4 above)
Comment (shuspes) [reply src/controllers/users.controller.js]: done

File: src/routes/users.route.js (hunk 7 above)
Comment (shuspes) [reply src/routes/users.route.js:8]: The idea was to create separate router instance for each route with its own set of middleware as an independent unit of application. router.use(express.json()); works only for user.router, and I need to add this line for any new router if this router is going to work with req.body. I don't wont to have json parser on the app layer.

File: src/routes/users.route.js (hunk 6 above)
Comment (shuspes) [reply src/routes/users.route.js]: I didn't see much sense in this, but I made changes)

File: src/controllers/users.controller.js (hunk 3 above)
Comment (shuspes) [reply src/controllers/users.controller.js]: it makes sense, updated

File: src/routes/users.route.js (hunk 5 above)
Comment (shuspes) [reply src/routes/users.route.js]: CustomError class was created to identify errors that are produced by my code

File: src/app.js (hunk 2 above)
Comment (shuspes) [reply src/app.js]: done

File: src/routes/users.route.js (hunk 7 above)
Comment (andrei-akula-epam) [reply src/routes/users.route.js:8]: but if you react API service, you're going to use json on all routes, don't you?

**Pull request merged**